contracts/covenant/src/contract.rs

```diff
@@ -0,0 +1,242 @@
+            msg: clock_instantiate_tx,
+            gas_limit: None,
+            reply_on: cosmwasm_std::ReplyOn::Always,
+        }
+    ))
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    deps.api
+        .debug(format!("WASMDEBUG: execute: received msg: {:?}", msg).as_str());
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::DepositorAddress {  } => Ok(to_binary(&COVENANT_DEPOSITOR.may_load(deps.storage)?)?),
+        QueryMsg::ClockAddress {  } => Ok(to_binary(&COVENANT_CLOCK.may_load(deps.storage)?)?),
+        QueryMsg::LpAddress {  } => Ok(to_binary(&COVENANT_LP.may_load(deps.storage)?)?),
+        QueryMsg::LsAddress {  } => Ok(to_binary(&COVENANT_LS.may_load(deps.storage)?)?),
+        QueryMsg::HolderAddress {  } => Ok(to_binary(&COVENANT_HOLDER.may_load(deps.storage)?)?),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    deps.api.debug("WASMDEBUG: migrate");
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractError> {
+
+    match msg.id {
+        CLOCK_REPLY_ID => handle_clock_reply(deps, env, msg),
+        HOLDER_REPLY_ID => handle_holder_reply(deps, env, msg),
+        LP_REPLY_ID => handle_lp_reply(deps, env, msg),
+        LS_REPLY_ID => handle_ls_reply(deps, env, msg),
+        _ => Err(ContractError::UnknownReplyId {}),
+    }
+}
+
+pub fn handle_clock_reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractError> {
+    deps.api.debug("WASMDEBUG: clock reply");
+
+    let parsed_data = parse_reply_instantiate_data(msg);
+    match parsed_data {
+        Ok(response) => {
+            // successful clock instantiation means we are ready to proceed with
+            // remaining instantiations
+            COVENANT_CLOCK.save(deps.storage, &response.contract_address)?;
+            
+            let holder_code = HOLDER_CODE.load(deps.storage)?;
+            let holder_data = HOLDER_INSTANTIATION_DATA.load(deps.storage)?;
+
+            let holder_instantiate_tx = CosmosMsg::Wasm(WasmMsg::Instantiate { 
+                admin: Some(env.contract.address.to_string()),
+                code_id: holder_code,
+                msg: to_binary(&holder_data)?,
+                funds: vec![],
+                label: "covenant-holder".to_string(),
+            });
+
+            Ok(Response::default().add_submessage(
+                SubMsg {
+                    id: HOLDER_REPLY_ID,
+                    msg: holder_instantiate_tx,
+                    gas_limit: None,
+                    reply_on: cosmwasm_std::ReplyOn::Always,
+                }
```

Comment:
I would use `on_success` here instead of always, because we don't really do anything when it failed, and just pass the message along, which will happen by default if the submsg fails.
It just makes it clear that we only expect success case, instead of both. 

You can use `SubMsg::reply_on_success(msg, id)` in that case.

Reply:
fair enough, convinced